Firestore/Swift/Source/PropertyWrapper/FirestoreQueryObservable.swift

```diff
@@ -0,0 +1,162 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import SwiftUI
+import FirebaseFirestore
+
+@available(iOS 14.0, *)
+@available(tvOS, unavailable)
+internal class FirestoreQueryObservable<T>: ObservableObject {
+  @Published var items: T
+
+  private let firestore = Firestore.firestore()
+  private var listener: ListenerRegistration?
+
+  private var setupListener: (() -> Void)!
+
+  internal var configuration: FirestoreQuery<T>.Configuration {
+    didSet {
+      removeListener()
+      setupListener()
+    }
+  }
+
+  init<U: Decodable>(configuration: FirestoreQuery<T>.Configuration) where T == [U] {
+    items = []
+    self.configuration = configuration
+    setupListener = createListener { [weak self] querySnapshot, error in
+      if let _ = error {
+        // in this scenario, we actively ignore the error
+        self?.items = []
+        return
+      }
+
+      guard let documents = querySnapshot?.documents else {
+        self?.items = []
+        return
+      }
+
+      self?.items = documents.compactMap { document in
+        try? document.data(as: U.self)
+      }
+    }
+
+    setupListener()
+  }
+
+  init<U: Decodable>(configuration: FirestoreQuery<T>.Configuration) where T == [Result<U, Error>] {
+    items = []
+    self.configuration = configuration
+    setupListener = createListener { [weak self] querySnapshot, error in
+      if let error = error {
+        self?.items = [.failure(error)]
+        return
+      }
+
+      guard let documents = querySnapshot?.documents else {
+        self?.items = []
+        return
+      }
+
+      self?.items = documents.compactMap { queryDocumentSnapshot in
+        Result {
+          // NOTE: The `try` handles the parse failure
+          // The `Optional` signature for the result is a
+          // workaround for when a document may not exist.
+          // - which they always do when looping over
+          // them in the `documents` property of a snapshot
+          try queryDocumentSnapshot.data(as: U.self)!
+        }
+      }
+    }
+
+    setupListener()
+  }
+
+  init<U: Decodable>(configuration: FirestoreQuery<T>.Configuration) where T == Result<[U], Error> {
+    items = .success([])
+    self.configuration = configuration
+    setupListener = createListener { [weak self] querySnapshot, error in
+      if let error = error {
+        self?.items = .failure(error)
+        return
+      }
+
+      guard let documents = querySnapshot?.documents else {
```

Comment:
Will a snapshot listener ever return a non-nil `querySnapshot` that has a `nil` `documents` reference?

@schmidt-sebastian @morganchen12

Reply:
Yes, since users can explicitly set nil documents. A non-nil query snapshot with a nil document indicates a successful fetch of nil, which, while confusing if not an intentional choice by the developer, is not an error.

Reply:
I am not suer I can follow this. A user can set `nil` values, but I don't see how documents can be nil here. It could very well be an empty list.

Reply:
@morganchen12 interesting - I did not know this. Can this be done in code only? This doesn't seem to be possible via the console.

@schmidt-sebastian are you saying that the Objective-C code never passes a `nil` value in `querySnapshot.documents`?

Reply:
Yes.